Chinese word become unrecognizable code using "addSchema(SCHEMA,scriptSources)" method

Using bellow cofig to start embedded-mysql:
````
MysqldConfig config = aMysqldConfig(v5_6_35)
                .withCharset(UTF8)
                .withPort(getPort())
                .withUser(USERNAME, PASSWORD)
                .withTimeout(2, TimeUnit.MINUTES)
                .withServerVariable("max_connect_errors", 666)
                .build();

EmbeddedMysql mysqld = anEmbeddedMysql(config, downloadConfig)
                .addSchema(SCHEMA,"db/001_init.sql")
                .start();
````
where db/001_init.sql:
````
create table t1 (col1 INTEGER NOT NULL,col2 VARCHAR(10));

INSERT INTO t1 values(10, "你好!");
````

after mysql start, "你好!" become unrecognizable code in table.

after research, probability fix:

add paramter to MysqlClient.java about line 54, after "--password=" with

**"--default-character-set=utf8"**.

@viliusl  I can see "
org.springframework.jdbc.CannotGetJdbcConnectionException: Could not get JDBC Connection; nested exception is java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user '你'@'localhost' (using password: YES))
Caused by: java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user '你'@'localhost' (using password: YES))
Caused by: java.sql.SQLException: Access denied for user '你'@'localhost' (using password: YES)
EmbeddedMysql instance should::support non-latin characters(com.wix.mysql.ExtendedCharsetTest)  Time "

seems test case (jdbc) using wrong user/password login mysql?
mybe need resolved #124 first?
I'm quite sure, your fix should work, I'm will try test case again.

--- a/wix-embedded-mysql/src/main/java/com/wix/mysql/EmbeddedMysql.java
+++ b/wix-embedded-mysql/src/main/java/com/wix/mysql/EmbeddedMysql.java
@@ -45,7 +45,7 @@ public class EmbeddedMysql {
 
         try {
             executable.start();
-            getClient(SCHEMA).executeCommands(
+            getClient(SCHEMA, mysqldConfig.getCharset()).executeCommands(
                     format("CREATE USER '%s'@'%%' IDENTIFIED BY '%s';", mysqldConfig.getUsername(), mysqldConfig.getPassword()));
         } catch (IOException e) {
             throw new RuntimeException(e);
@@ -69,19 +69,20 @@ public class EmbeddedMysql {
         addSchema(config);
     }
 
-    public void dropSchema(final SchemaConfig config) {
-        getClient(SystemDefaults.SCHEMA).executeCommands(format("DROP DATABASE %s", config.getName()));
+    public void dropSchema(final SchemaConfig schema) {
+        Charset effectiveCharset = or(schema.getCharset(), config.getCharset());
+        getClient(SystemDefaults.SCHEMA, effectiveCharset).executeCommands(format("DROP DATABASE %s", schema.getName()));
     }
 
     public EmbeddedMysql addSchema(final SchemaConfig schema) {
         Charset effectiveCharset = or(schema.getCharset(), config.getCharset());
 
-        getClient(SystemDefaults.SCHEMA).executeCommands(
+        getClient(SystemDefaults.SCHEMA, effectiveCharset).executeCommands(
                 format("CREATE DATABASE %s CHARACTER SET = %s COLLATE = %s;",
                         schema.getName(), effectiveCharset.getCharset(), effectiveCharset.getCollate()),
                 format("GRANT ALL ON %s.* TO '%s'@'%%';", schema.getName(), config.getUsername()));
 
-        MysqlClient client = getClient(schema.getName());
+        MysqlClient client = getClient(schema.getName(), effectiveCharset);
         client.executeScripts(schema.getScripts());
 
         return this;
@@ -93,8 +94,8 @@ public class EmbeddedMysql {
         }
     }
 
-    private MysqlClient getClient(final String schemaName) {
-        return new MysqlClient(config, executable, schemaName);
+    private MysqlClient getClient(final String schemaName, final Charset charset) {
+        return new MysqlClient(config, executable, schemaName, charset);
     }
 
     public static Builder anEmbeddedMysql(

--- a/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
+++ b/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
@@ -63,7 +63,7 @@ class MysqlClient {
                     format("--port=%s", config.getPort()),
                     schemaName}).start();
 
-            IOUtils.copy(new StringReader(sql), p.getOutputStream(), java.nio.charset.Charset.forName("UTF-8"));
+            IOUtils.copy(new StringReader(sql), p.getOutputStream(), java.nio.charset.Charset.defaultCharset());
             p.getOutputStream().close();
 
             if (p.waitFor() != 0) {

--- a/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
+++ b/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
@@ -63,7 +63,7 @@ class MysqlClient {
                     format("--port=%s", config.getPort()),
                     schemaName}).start();
 
-            IOUtils.copy(new StringReader(sql), p.getOutputStream(), java.nio.charset.Charset.defaultCharset());
+            IOUtils.copy(new StringReader(sql), p.getOutputStream(), java.nio.charset.Charset.forName("UTF-8"));
             p.getOutputStream().close();
 
             if (p.waitFor() != 0) {

--- a/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
+++ b/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
@@ -58,7 +58,7 @@ class MysqlClient {
                     "--protocol=tcp",
                     "--host=localhost",
                     "--password=",
-                    format("--default-character-set=$s", effectiveCharset.getCharset()),
+                    format("--default-character-set=%s", effectiveCharset.getCharset()),
                     format("--user=%s", SystemDefaults.USERNAME),
                     format("--port=%s", config.getPort()),
                     schemaName}).start();

--- a/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
+++ b/wix-embedded-mysql/src/main/java/com/wix/mysql/MysqlClient.java
@@ -1,5 +1,6 @@
 package com.wix.mysql;
 
+import com.wix.mysql.config.Charset;
 import com.wix.mysql.config.MysqldConfig;
 import com.wix.mysql.config.MysqldConfig.SystemDefaults;
 import com.wix.mysql.exceptions.CommandFailedException;
@@ -8,7 +9,6 @@ import org.apache.commons.io.IOUtils;
 
 import java.io.IOException;
 import java.io.StringReader;
-import java.nio.charset.Charset;
 import java.nio.file.Paths;
 import java.util.List;
 
@@ -21,11 +21,17 @@ class MysqlClient {
     private final MysqldConfig config;
     private final MysqldExecutable executable;
     private final String schemaName;
+    private final Charset effectiveCharset;
 
-    public MysqlClient(final MysqldConfig config, final MysqldExecutable executable, final String schemaName) {
+    public MysqlClient(
+            final MysqldConfig config,
+            final MysqldExecutable executable,
+            final String schemaName,
+            final Charset effectiveCharset) {
         this.config = config;
         this.executable = executable;
         this.schemaName = schemaName;
+        this.effectiveCharset = effectiveCharset;
     }
 
     void executeScripts(final List<SqlScriptSource> sqls) {
@@ -52,11 +58,12 @@ class MysqlClient {
                     "--protocol=tcp",
                     "--host=localhost",
                     "--password=",
+                    format("--default-character-set=$s", effectiveCharset.getCharset()),
                     format("--user=%s", SystemDefaults.USERNAME),
                     format("--port=%s", config.getPort()),
                     schemaName}).start();
 
-            IOUtils.copy(new StringReader(sql), p.getOutputStream(), Charset.defaultCharset());
+            IOUtils.copy(new StringReader(sql), p.getOutputStream(), java.nio.charset.Charset.defaultCharset());
             p.getOutputStream().close();
 
             if (p.waitFor() != 0) {

--- a/wix-embedded-mysql/src/test/resources/db/004_update3.sql
+++ b/wix-embedded-mysql/src/test/resources/db/004_update3.sql
@@ -1,0 +1,2 @@
+create table t1 (col1 INTEGER, col2 VARCHAR(10));
+INSERT INTO t1 values(1, '你好!');

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -11,7 +11,7 @@ class ExtendedCharsetTest extends IntegrationTest {
 
   "EmbeddedMysql instance" should {
 
-    "support non-latin characters" in {
+    "support non-latin characters in login/password" in {
       todo
       val config = testConfigBuilder
         .withCharset(Charset.UTF8MB4)

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -1,0 +1,39 @@
+package com.wix.mysql
+
+import com.wix.mysql.EmbeddedMysql._
+import com.wix.mysql.ScriptResolver.classPathScript
+import com.wix.mysql.config.MysqldConfig.SystemDefaults
+import com.wix.mysql.support.IntegrationTest
+import com.wix.mysql.support.IntegrationTest._
+
+class ExtendedCharsetTest extends IntegrationTest {
+
+  "EmbeddedMysql instance" should {
+
+    "support non-latin characters" in {
+      val config = testConfigBuilder
+        .withUser("你", "好").build
+
+      val mysqld = start(anEmbeddedMysql(config))
+
+      mysqld must beAvailableOn(3310, "你", "好", SystemDefaults.SCHEMA)
+    }
+
+    "support migration from file with extended charset" in {
+      val mysqld = start(anEmbeddedMysql(testConfigBuilder.build)
+        .addSchema("aSchema", classPathScript("/db/004_update3.sql")))
+
+      aSelect[java.lang.String](mysqld, onSchema = "aSchema", sql = "select col2 from t1 where col1 = 1;") mustEqual "你好!"
+    }
+
+    "support inline migration with extended charset" in {
+      val mysqld = start(anEmbeddedMysql(testConfigBuilder.build)
+        .addSchema(
+          "aSchema",
+          aMigrationWith("create table t1 (col1 INTEGER, col2 VARCHAR(10));\nINSERT INTO t1 values(1, '你好!');")))
+
+      aSelect[java.lang.String](mysqld, onSchema = "aSchema", sql = "select col2 from t1 where col1 = 1;") mustEqual "你好!"
+    }
+
+  }
+}

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -12,6 +12,7 @@ class ExtendedCharsetTest extends IntegrationTest {
   "EmbeddedMysql instance" should {
 
     "support non-latin characters" in {
+      todo
       val config = testConfigBuilder
         .withCharset(Charset.UTF8MB4)
         .withUser("你", "好").build

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -13,6 +13,7 @@ class ExtendedCharsetTest extends IntegrationTest {
 
 //    "support non-latin characters" in {
 //      val config = testConfigBuilder
+//        .withCharset(Charset.UTF8MB4)
 //        .withUser("你", "好").build
 //
 //      val mysqld = start(anEmbeddedMysql(config))

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -2,6 +2,7 @@ package com.wix.mysql
 
 import com.wix.mysql.EmbeddedMysql._
 import com.wix.mysql.ScriptResolver.classPathScript
+import com.wix.mysql.config.Charset
 import com.wix.mysql.config.MysqldConfig.SystemDefaults
 import com.wix.mysql.support.IntegrationTest
 import com.wix.mysql.support.IntegrationTest._
@@ -20,14 +21,14 @@ class ExtendedCharsetTest extends IntegrationTest {
     }
 
     "support migration from file with extended charset" in {
-      val mysqld = start(anEmbeddedMysql(testConfigBuilder.build)
+      val mysqld = start(anEmbeddedMysql(testConfigBuilder.withCharset(Charset.UTF8MB4).build)
         .addSchema("aSchema", classPathScript("/db/004_update3.sql")))
 
       aSelect[java.lang.String](mysqld, onSchema = "aSchema", sql = "select col2 from t1 where col1 = 1;") mustEqual "你好!"
     }
 
     "support inline migration with extended charset" in {
-      val mysqld = start(anEmbeddedMysql(testConfigBuilder.build)
+      val mysqld = start(anEmbeddedMysql(testConfigBuilder.withCharset(Charset.UTF8MB4).build)
         .addSchema(
           "aSchema",
           aMigrationWith("create table t1 (col1 INTEGER, col2 VARCHAR(10));\nINSERT INTO t1 values(1, '你好!');")))

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -11,7 +11,7 @@ class ExtendedCharsetTest extends IntegrationTest {
 
   "EmbeddedMysql instance" should {
 
-    "support non-latin characters for login/password" in {
+    "support non-latin characters" in {
       val config = testConfigBuilder
         .withCharset(Charset.UTF8MB4)
         .withUser("你", "好").build

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -11,15 +11,15 @@ class ExtendedCharsetTest extends IntegrationTest {
 
   "EmbeddedMysql instance" should {
 
-//    "support non-latin characters" in {
-//      val config = testConfigBuilder
-//        .withCharset(Charset.UTF8MB4)
-//        .withUser("你", "好").build
-//
-//      val mysqld = start(anEmbeddedMysql(config))
-//
-//      mysqld must beAvailableOn(3310, "你", "好", SystemDefaults.SCHEMA)
-//    }
+    "support non-latin characters" in {
+      val config = testConfigBuilder
+        .withCharset(Charset.UTF8MB4)
+        .withUser("你", "好").build
+
+      val mysqld = start(anEmbeddedMysql(config))
+
+      mysqld must beAvailableOn(3310, "你", "好", SystemDefaults.SCHEMA)
+    }
 
     "support migration from file with extended charset" in {
       val mysqld = start(anEmbeddedMysql(testConfigBuilder.withCharset(Charset.UTF8MB4).build)

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -11,14 +11,14 @@ class ExtendedCharsetTest extends IntegrationTest {
 
   "EmbeddedMysql instance" should {
 
-    "support non-latin characters" in {
-      val config = testConfigBuilder
-        .withUser("你", "好").build
-
-      val mysqld = start(anEmbeddedMysql(config))
-
-      mysqld must beAvailableOn(3310, "你", "好", SystemDefaults.SCHEMA)
-    }
+//    "support non-latin characters" in {
+//      val config = testConfigBuilder
+//        .withUser("你", "好").build
+//
+//      val mysqld = start(anEmbeddedMysql(config))
+//
+//      mysqld must beAvailableOn(3310, "你", "好", SystemDefaults.SCHEMA)
+//    }
 
     "support migration from file with extended charset" in {
       val mysqld = start(anEmbeddedMysql(testConfigBuilder.withCharset(Charset.UTF8MB4).build)

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ExtendedCharsetTest.scala
@@ -11,7 +11,7 @@ class ExtendedCharsetTest extends IntegrationTest {
 
   "EmbeddedMysql instance" should {
 
-    "support non-latin characters" in {
+    "support non-latin characters for login/password" in {
       val config = testConfigBuilder
         .withCharset(Charset.UTF8MB4)
         .withUser("你", "好").build

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/MysqlTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/MysqlTest.scala
@@ -1,6 +1,7 @@
 package com.wix.mysql
 
 import com.wix.mysql.EmbeddedMysql.anEmbeddedMysql
+import com.wix.mysql.config.Charset
 import com.wix.mysql.exceptions.CommandFailedException
 import com.wix.mysql.support.IntegrationTest
 import com.wix.mysql.support.IntegrationTest.targetTestVersion
@@ -9,7 +10,7 @@ class MysqlTest extends IntegrationTest {
 
   "mysql should emit exception info with message from 'mysql' command output'" in {
     val mysqld = start(anEmbeddedMysql(targetTestVersion))
-    val mysql = new MysqlClient(mysqld.getConfig, mysqld.executable, "information_schema")
+    val mysql = new MysqlClient(mysqld.getConfig, mysqld.executable, "information_schema", Charset.UTF8MB4)
 
     mysql.executeCommands("sele qwe from zz;") must throwA[CommandFailedException].like {
       case e: CommandFailedException => e.getMessage must contain(

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/ScriptResolverTest.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/ScriptResolverTest.scala
@@ -12,6 +12,7 @@ class ScriptResolverTest extends SpecificationWithJUnit with FileMatchers {
   val contentsOf001Init = "create table t1 "
   val contentsOf002Update = "create table t2 "
   val contentsOf003Update = "create table t3 "
+  val contentsOf004Update = "create table t1 "
   val contentsOf001InitInJar = "create table t1_jar"
   val contentsOf002UpdateInJar = "create table t2_jar"
 
@@ -33,11 +34,19 @@ class ScriptResolverTest extends SpecificationWithJUnit with FileMatchers {
   "classPathFiles" should {
 
     "resolve multiple classPath files" in {
-      classPathFiles("/db/*.sql") must containScripts(contentsOf001Init, contentsOf002Update, contentsOf003Update)
+      classPathFiles("/db/*.sql") must containScripts(
+        contentsOf001Init,
+        contentsOf002Update,
+        contentsOf003Update,
+        contentsOf004Update)
     }
 
     "resolve multiple classPath files without preceding '/'" in {
-      classPathFiles("db/*.sql") must containScripts(contentsOf001Init, contentsOf002Update, contentsOf003Update)
+      classPathFiles("db/*.sql") must containScripts(
+        contentsOf001Init,
+        contentsOf002Update,
+        contentsOf003Update,
+        contentsOf004Update)
     }
 
     "throw a ScriptResolutionException if no classPathFiles are found" in {
@@ -71,11 +80,19 @@ class ScriptResolverTest extends SpecificationWithJUnit with FileMatchers {
   "classPathScripts" should {
 
     "resolve multiple classPath files" in {
-      classPathScripts("/db/*.sql") must containScripts(contentsOf001Init, contentsOf002Update, contentsOf003Update)
+      classPathScripts("/db/*.sql") must containScripts(
+        contentsOf001Init,
+        contentsOf002Update,
+        contentsOf003Update,
+        contentsOf004Update)
     }
 
     "resolve multiple classPath files without preceding '/'" in {
-      classPathScripts("db/*.sql") must containScripts(contentsOf001Init, contentsOf002Update, contentsOf003Update)
+      classPathScripts("db/*.sql") must containScripts(
+        contentsOf001Init,
+        contentsOf002Update,
+        contentsOf003Update,
+        contentsOf004Update)
     }
 
     "resolve multiple classPath scripts within jar in classpath" in {

--- a/wix-embedded-mysql/src/test/scala/com/wix/mysql/support/TestResourceSupport.scala
+++ b/wix-embedded-mysql/src/test/scala/com/wix/mysql/support/TestResourceSupport.scala
@@ -1,6 +1,7 @@
 package com.wix.mysql.support
 
 import java.io.{File => JFile}
+import java.nio.charset.Charset
 import java.util.UUID
 
 import org.apache.commons.io.FileUtils
@@ -12,7 +13,7 @@ trait TestResourceSupport {
   def createTempFile(content: String): JFile = {
     testResourceDirExists()
     val file = new JFile(s"$targetDir/classes/${UUID.randomUUID}")
-    FileUtils.writeStringToFile(file, content)
+    FileUtils.writeStringToFile(file, content, Charset.forName("UTF-8"))
     file
   }
 

